pkg/proxyserver/controllers/service_resolver_controller.go

```diff
@@ -0,0 +1,131 @@
+	r := &ServiceResolverReconciler{
+		Client: mgr.GetClient(),
+	}
+	return r.SetupWithManager(mgr)
+}
+
+func (c *ServiceResolverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&proxyv1alpha1.ManagedProxyServiceResolver{}).
+		Watches(
+			&source.Kind{
+				Type: &proxyv1alpha1.ManagedProxyServiceResolver{},
+			},
+			&eventhandler.ProxyServiceResolverHandler{},
+		).
+		Watches(
+			&source.Kind{
+				Type: &clusterv1beta1.ManagedClusterSet{},
+			},
+			&eventhandler.ClustersetHandler{
+				Client: mgr.GetClient(),
+			},
+		).
+		Complete(c)
+}
+
+func (c *ServiceResolverReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log.Info("Reconciling ServiceResolver", "name", req.Name)
+	// refreshing service resolvers status
+	return ctrl.Result{}, c.refreshManageProxyServiceResolversStatus()
+}
+
+func (c *ServiceResolverReconciler) refreshManageProxyServiceResolversStatus() error {
+	// list ManagedProxyServiceResolvers
+	resolvers := &proxyv1alpha1.ManagedProxyServiceResolverList{}
+	if err := c.Client.List(context.TODO(), resolvers); err != nil {
+		return err
+	}
+
+	var errs []error
+
+	// for each resolver, get it's managedclusterset
+	for i := range resolvers.Items {
+		resolver := resolvers.Items[i]
+		var expectingCondition metav1.Condition
+		currentConditions := resolver.Status.DeepCopy().Conditions
+
+		// Currently, managedclusterseletor only support clusterset type, and serviceselector only support serviceRef type.
+		if !util.IsServiceResolverLegal(&resolver) {
+			expectingCondition = metav1.Condition{
+				Type:   proxyv1alpha1.ConditionTypeServiceResolverAvaliable,
+				Status: metav1.ConditionFalse,
+				Reason: "ManagedProxyServiceResolverNotLegal",
+			}
+		} else {
+			// get managedclusterset
+			managedClusterSet := &clusterv1beta1.ManagedClusterSet{}
+			if err := c.Client.Get(context.TODO(), types.NamespacedName{
+				Name: resolver.Spec.ManagedClusterSelector.ManagedClusterSet.Name,
+			}, managedClusterSet); err != nil {
+				if apierrors.IsNotFound(err) {
+					expectingCondition = metav1.Condition{
+						Type:   proxyv1alpha1.ConditionTypeServiceResolverAvaliable,
+						Status: metav1.ConditionFalse,
+						Reason: "ManagedClusterSetNotExisted",
+					}
+				} else {
+					return err
+				}
+			} else {
+				if !managedClusterSet.DeletionTimestamp.IsZero() {
+					expectingCondition = metav1.Condition{
+						Type:   proxyv1alpha1.ConditionTypeServiceResolverAvaliable,
+						Status: metav1.ConditionFalse,
+						Reason: "ManagedClusterSetDeleting",
+					}
+				} else {
+					expectingCondition = metav1.Condition{
+						Type:   proxyv1alpha1.ConditionTypeServiceResolverAvaliable,
+						Status: metav1.ConditionTrue,
+						Reason: "ManagedProxyServiceResolverAvaliable",
+					}
+				}
+			}
+		}
+
+		// update status; now only consider one condition.
+		if len(currentConditions) != 0 {
```

Comment:
you may need the meta.FindStatusCondition

Reply:
Updated.